src/utils/polygonUtils.cpp

```diff
@@ -1347,4 +1347,56 @@ double PolygonUtils::relativeHammingDistance(const Polygons& poly_a, const Polyg
     return hamming_distance / total_area;
 }
 
+void PolygonUtils::makeCircle(Point mid, coord_t radius, Polygons& ret, float a_step)
+{
+    PolygonRef circle = ret.newPoly();
+    for (float a = 0; a < 2 * M_PI; a += a_step)
+    {
+        circle.emplace_back(mid + Point(radius * cos(a), radius * sin(a)));
+    }
+}
+
+
+Polygons PolygonUtils::connect(const Polygons& input)
+{
+    Polygons ret;
+    std::vector<PolygonsPart> parts = input.splitIntoParts(true);
+    for (PolygonsPart& part : parts)
+    {
+        PolygonRef outline = part.outerPolygon();
+        for (size_t hole_idx = 1; hole_idx < part.size(); hole_idx++)
+        {
+            PolygonRef hole = part[hole_idx];
+            Point hole_point = hole[0];
+            hole.add(hole_point);
+            // find where the scanline passes the Y
+            size_t best_segment_to_idx = 0;
+            coord_t best_dist = std::numeric_limits<coord_t>::max();
```

Comment:
We have utility functions to find where polygons are closest, right?

Reply:
True, but this routine also prevents creating a line that self-intersects with the hole outline, which would be impossible then.